### ISSUE TITLE
btrfs/104 lockdep warning (rwsem) &fs_info->commit_root_sem btrfs-root-00 

### ISSUE THREAD

**Issue by josefbacik** [btrfs/fstests#43]
```
[   59.757442] ======================================================
[   59.757854] WARNING: possible circular locking dependency detected
[   59.758277] 5.9.0+ #101 Not tainted
[   59.758513] ------------------------------------------------------
[   59.758929] btrfs-cleaner/3445 is trying to acquire lock:
[   59.759289] ffff89dbec39ab48 (btrfs-root-00){++++}-{3:3}, at: __btrfs_tree_read_lock+0x32/0x170
[   59.759876]
[   59.759876] but task is already holding lock:
[   59.760271] ffff89dbeaf28a88 (&fs_info->commit_root_sem){++++}-{3:3}, at: btrfs_find_all_roots+0x41/0x80
[   59.760900]
[   59.760900] which lock already depends on the new lock.
[   59.760900]
[   59.761448]
[   59.761448] the existing dependency chain (in reverse order) is:
[   59.761950]
[   59.761950] -> #2 (&fs_info->commit_root_sem){++++}-{3:3}:
[   59.762416]        down_write+0x3d/0x70
[   59.762678]        btrfs_cache_block_group+0x2d5/0x510
[   59.763024]        find_free_extent+0xb6e/0x12f0
[   59.763335]        btrfs_reserve_extent+0xb3/0x1b0
[   59.763657]        btrfs_alloc_tree_block+0xb1/0x330
[   59.763991]        alloc_tree_block_no_bg_flush+0x4f/0x60
[   59.764351]        __btrfs_cow_block+0x11d/0x580
[   59.764660]        btrfs_cow_block+0x10c/0x220
[   59.764969]        commit_cowonly_roots+0x47/0x2e0
[   59.765289]        btrfs_commit_transaction+0x595/0xbd0
[   59.765641]        sync_filesystem+0x74/0x90
[   59.765929]        generic_shutdown_super+0x22/0x100
[   59.766266]        kill_anon_super+0x14/0x30
[   59.766550]        btrfs_kill_super+0x12/0x20
[   59.766842]        deactivate_locked_super+0x36/0xa0
[   59.767176]        cleanup_mnt+0x12d/0x190
[   59.767452]        task_work_run+0x5c/0xa0
[   59.767730]        exit_to_user_mode_prepare+0x1df/0x200
[   59.768088]        syscall_exit_to_user_mode+0x54/0x280
[   59.768442]        entry_SYSCALL_64_after_hwframe+0x44/0xa9
[   59.768814]
[   59.768814] -> #1 (&space_info->groups_sem){++++}-{3:3}:
[   59.769269]        down_read+0x40/0x130
[   59.769529]        find_free_extent+0x2ed/0x12f0
[   59.769840]        btrfs_reserve_extent+0xb3/0x1b0
[   59.770165]        btrfs_alloc_tree_block+0xb1/0x330
[   59.770494]        alloc_tree_block_no_bg_flush+0x4f/0x60
[   59.770850]        __btrfs_cow_block+0x11d/0x580
[   59.771162]        btrfs_cow_block+0x10c/0x220
[   59.771463]        commit_cowonly_roots+0x47/0x2e0
[   59.771785]        btrfs_commit_transaction+0x595/0xbd0
[   59.772137]        sync_filesystem+0x74/0x90
[   59.772426]        generic_shutdown_super+0x22/0x100
[   59.772760]        kill_anon_super+0x14/0x30
[   59.773051]        btrfs_kill_super+0x12/0x20
[   59.773345]        deactivate_locked_super+0x36/0xa0
[   59.773679]        cleanup_mnt+0x12d/0x190
[   59.773958]        task_work_run+0x5c/0xa0
[   59.774234]        exit_to_user_mode_prepare+0x1df/0x200
[   59.774591]        syscall_exit_to_user_mode+0x54/0x280
[   59.774938]        entry_SYSCALL_64_after_hwframe+0x44/0xa9
[   59.775314]
[   59.775314] -> #0 (btrfs-root-00){++++}-{3:3}:
[   59.775709]        __lock_acquire+0x1167/0x2150
[   59.776018]        lock_acquire+0xb9/0x3d0
[   59.776298]        down_read_nested+0x43/0x130
[   59.776597]        __btrfs_tree_read_lock+0x32/0x170
[   59.776931]        __btrfs_read_lock_root_node+0x3a/0x50
[   59.777288]        btrfs_search_slot+0x614/0x9d0
[   59.777590]        btrfs_find_root+0x35/0x1b0
[   59.777885]        btrfs_read_tree_root+0x61/0x120
[   59.778210]        btrfs_get_root_ref+0x14b/0x600
[   59.778526]        find_parent_nodes+0x3e6/0x1b30
[   59.778845]        btrfs_find_all_roots_safe+0xb4/0x130
[   59.779196]        btrfs_find_all_roots+0x60/0x80
[   59.779510]        btrfs_qgroup_trace_extent_post+0x27/0x40
[   59.779882]        btrfs_add_delayed_data_ref+0x3fd/0x460
[   59.780245]        btrfs_free_extent+0x42/0x100
[   59.780554]        __btrfs_mod_ref+0x1d7/0x2f0
[   59.780848]        walk_up_proc+0x11c/0x400
[   59.781133]        walk_up_tree+0xf0/0x180
[   59.781408]        btrfs_drop_snapshot+0x1c7/0x780
[   59.781728]        btrfs_clean_one_deleted_snapshot+0xfb/0x110
[   59.782114]        cleaner_kthread+0xd4/0x140
[   59.782404]        kthread+0x13a/0x150
[   59.782657]        ret_from_fork+0x1f/0x30
[   59.782931]
[   59.782931] other info that might help us debug this:
[   59.782931]
[   59.783463] Chain exists of:
[   59.783463]   btrfs-root-00 --> &space_info->groups_sem --> &fs_info->commit_root_sem
[   59.783463]
[   59.784266]  Possible unsafe locking scenario:
[   59.784266]
[   59.784663]        CPU0                    CPU1
[   59.784971]        ----                    ----
[   59.785282]   lock(&fs_info->commit_root_sem);
[   59.785580]                                lock(&space_info->groups_sem);
[   59.786032]                                lock(&fs_info->commit_root_sem);
[   59.786492]   lock(btrfs-root-00);
[   59.786725]
[   59.786725]  *** DEADLOCK ***
[   59.786725]
[   59.787122] 3 locks held by btrfs-cleaner/3445:
[   59.787424]  #0: ffff89dbeaf28838 (&fs_info->cleaner_mutex){+.+.}-{3:3}, at: cleaner_kthread+0x6e/0x140
[   59.788047]  #1: ffff89dbeb6c7640 (sb_internal){.+.+}-{0:0}, at: start_transaction+0x40b/0x5c0
[   59.788616]  #2: ffff89dbeaf28a88 (&fs_info->commit_root_sem){++++}-{3:3}, at: btrfs_find_all_roots+0x41/0x80
[   59.789277]
[   59.789277] stack backtrace:
[   59.789569] CPU: 0 PID: 3445 Comm: btrfs-cleaner Not tainted 5.9.0+ #101
[   59.790017] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS 1.13.0-2.fc32 04/01/2014
[   59.790598] Call Trace:
[   59.790771]  dump_stack+0x8b/0xb0
[   59.791000]  check_noncircular+0xcf/0xf0
[   59.791268]  __lock_acquire+0x1167/0x2150
[   59.791540]  ? __bfs+0x42/0x210
[   59.791755]  lock_acquire+0xb9/0x3d0
[   59.792002]  ? __btrfs_tree_read_lock+0x32/0x170
[   59.792316]  down_read_nested+0x43/0x130
[   59.792581]  ? __btrfs_tree_read_lock+0x32/0x170
[   59.792890]  __btrfs_tree_read_lock+0x32/0x170
[   59.793195]  __btrfs_read_lock_root_node+0x3a/0x50
[   59.793517]  btrfs_search_slot+0x614/0x9d0
[   59.793794]  ? find_held_lock+0x2b/0x80
[   59.794060]  btrfs_find_root+0x35/0x1b0
[   59.794319]  ? do_raw_spin_unlock+0x4b/0xa0
[   59.794603]  btrfs_read_tree_root+0x61/0x120
[   59.794892]  btrfs_get_root_ref+0x14b/0x600
[   59.795179]  find_parent_nodes+0x3e6/0x1b30
[   59.795465]  btrfs_find_all_roots_safe+0xb4/0x130
[   59.795784]  btrfs_find_all_roots+0x60/0x80
[   59.796071]  btrfs_qgroup_trace_extent_post+0x27/0x40
[   59.796413]  btrfs_add_delayed_data_ref+0x3fd/0x460
[   59.796744]  btrfs_free_extent+0x42/0x100
[   59.797019]  __btrfs_mod_ref+0x1d7/0x2f0
[   59.797287]  walk_up_proc+0x11c/0x400
[   59.797536]  walk_up_tree+0xf0/0x180
[   59.797781]  btrfs_drop_snapshot+0x1c7/0x780
[   59.798073]  ? btrfs_clean_one_deleted_snapshot+0x73/0x110
[   59.798441]  btrfs_clean_one_deleted_snapshot+0xfb/0x110
[   59.798798]  cleaner_kthread+0xd4/0x140
[   59.799059]  ? btrfs_alloc_root+0x50/0x50
[   59.799332]  kthread+0x13a/0x150
[   59.799556]  ? kthread_create_worker_on_cpu+0x40/0x40
[   59.799894]  ret_from_fork+0x1f/0x30
```

**Comment by josefbacik** [comment]
this is misc-next which has my rwsem and my fix for #42

**Comment by kdave** [comment]
Fixed by https://lore.kernel.org/linux-btrfs/2402f8ee46e0e080f9c22115b7ba3a02962ded4e.1603137558.git.josef@toxicpanda.com/
